algebird-core/src/main/scala/com/twitter/algebird/HyperLogLog.scala

```diff
@@ -79,34 +80,65 @@ object HyperLogLog {
   @inline
   def twopow(i: Int): Double = java.lang.Math.pow(2.0, i)
 
+  @deprecated("this is no longer used", since = "1.12.3")
+  def j(bsl: BitSetLite, bits: Int): Int =
+    j(bsl.in, bits)
+
   /**
-   * the value 'j' is equal to <w_0, w_1 ... w_(bits-1)>
-   *  TODO: We could read in a byte at a time.
+   *
    */
-  def j(bsl: BitSetLite, bits: Int): Int = {
-    @annotation.tailrec
-    def loop(pos: Int, accum: Int): Int = {
-      if (pos >= bits) {
-        accum
-      } else if (bsl.contains(pos)) {
-        loop(pos + 1, accum + (1 << pos))
-      } else {
-        loop(pos + 1, accum)
+  def j(bytes: Array[Byte], bits: Int): Int = {
+    var i = 0
+    var sum = 0
+    var need = bits
+    while (i < bytes.length && need >= 0) {
+      val byte = bytes(i) & 0xff
+      val limit = java.lang.Math.min(8, need)
+      var j = 0
+      while (j < limit) {
+        sum += ((byte >>> (7 - j)) & 1) << (i * 8 + j)
+        j += 1
       }
+      need -= 8
+      i += 1
     }
-    loop(0, 0)
+    sum
   }
 
+  @deprecated("this is no longer used", since = "1.12.3")
+  def rhoW(bsl: BitSetLite, bits: Int): Byte =
+    rhoW(bsl.in, bits)
+
   /**
-   * The value 'w' is equal to <w_bits ... w_n>. The function rho counts the number of leading
-   *  zeroes in 'w'. We can calculate rho(w) at once with the method rhoW.
+   * The value 'w' is represented as a bitset (encoding in
+   * `bytes`). This function counds the number of leading zeros in 'w'.
+   *
+   * Each byte is treated as a set of bits (little-endian). That is,
+   * the one bit represents the first value, then the two bit, then
+   * four, and so on.
+   *
+   * We treat the leading `bits` bits as if they were instead a single
+   * zero bit.
    */
-  def rhoW(bsl: BitSetLite, bits: Int): Byte = {
-    @annotation.tailrec
-    def loop(pos: Int, zeros: Int): Int =
-      if (bsl.contains(pos)) zeros
-      else loop(pos + 1, zeros + 1)
-    loop(bits, 1).toByte
+  def rhoW(bytes: Array[Byte], bits: Int): Byte = {
+    var i = bits / 8 // tracks the position in bytes
+    var j = 7 - (bits % 8) // tracks the bit position in bytes(i)
+    var zeros = 1 // start with a single zero
+    while (i < bytes.length) {
+      while (j >= 0) {
+        if (((bytes(i) >>> j) & 1) == 1) return zeros.toByte
```

Comment:
nit: feel free to ignore, but I think having `return` on its own line is worth it here. Mainly, I saw the 0.toByte below, figured there was a `return` somewhere, but wasn't able to easily locate the return statement in this code.